docs/ocean/design_docs/vert-lagrang-remap/vert-lagrang-remap.rst

```diff
@@ -0,0 +1,524 @@
+
+For the regridding step, layer thicknesses are set according to the target 
+grid, conserving volume:
+
+.. math::
+
+   h_k^{t+1} = h_k^{target}
+   
+   \sum_{k=1}^{kmax}h_k^{t+1} = \sum_{k=1}^{kmax}h_k^{lg}
+
+
+For the remapping step, velocities (edge-normal) and scalars are remapped to 
+the target grid, conserving volume flux and scalar concentration:
+
+.. math::
+
+   \sum_{k=1}^{kmax} u_k^{t+1} h_k^{t+1} = \sum_{k=1}^{kmax} u_k^{lg} h_k^{lg}
+   
+   \sum_{k=1}^{kmax} \phi_k^{t+1} h_k^{t+1} = \sum_{k=1}^{kmax} \phi_k^{lg} h_k^{lg}
+
+The vertical velocity across layer interfaces may be computed anytime after 
+regridding. It can be computed as 
+
+.. math::
+
+   w = - \nabla \cdot (h_k \mathbf{u}_k) - (h_k^{t+1} - h_k^t)/dt
+
+or
+
+.. math::
+
+   w = (h_k^{t+1} - h_k^lg)/dt
+
+The choice between the two is discussed in the Implementation section.
+
+
+Implementation
+--------------
+
+Implementation: Ability to perform vertical Lagrangian-remapping
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2020/12/15
+
+Contributors: Darren Engwirda, Xylar Asay-Davis, Carolyn Begeman
+
+Namelist options:
+
+- To turn VLR on/off: 
+  :code:`ALE_vertical_grid, config_vert_lagrangian_remap = .true. or .false.`
+- *Something related to target grid, for now just z_initial*
+
+Namelist options specific to PPR are discussed later.
+
+Lagrangian step:
+
+The solution for prognostic variables in RK4 and split-explicit remains
+largely the same. The main difference is that the vertical velocity through 
+the top of layers is set to zero in the routine 
+:code:`ocn_vert_transport_velocity_top`. This is similar to what is done when 
+:code:`config_vert_coord_movement` is :code:`impermeable_interfaces`, except 
+rather than exit the routine, we proceed with computations needed for the z-star and
+z-tilde coordinate choices.
+
+*Other modifications to ocn_vert_transport_velocity_top are not yet determined.*
+We will likely need to bypass the :code:`ocn_ALE_thickness` call in 
+:code:`ocn_vert_transport_velocity_top` so that the adjustments of layer 
+thickness for SSH perturbations occur during the regridding step.
+
+
+Regridding step(s):
+
+#. :math:`z_k^1`, the depth of the top of the layer, is determined based on 
+   an analytical expression for the grid. 
+   The simplest case is constant z-levels, :math:`z_k^1 = z_k^{init}`.
+   Since :math:`z_k^1` can be a function of the ocean state (e.g., :math:`\rho` 
+   for isopycnal coordinates, regridding doesn't begin until after the solution 
+   for prognostic variables.
+#. Superimpose SSH perturbations according to one of the existing depth-
+   dependent functions, :math:`z_k^2 = z_k^1 + c(z) \: \eta`. As in 
+   :code:`ocn_ALE_thickness`, layer thicknesses are adjusted from the seafloor 
+   upwards. Ideally, there is a single function that is used for both ALE
+   implementations, with and without VLR.
```

Comment:
I'm a bit confused by this notation.  Since `k` is the vertical index, shouldn't this be `z_1`, `z_2`, etc. instead of `z_k^1`, `z_k^2`, etc?

Separate from that confusion, I'm not clear on why the top layer is handled differently than subsequent layers.  I would expect one would have a list of layer-interface locations `z_k^\textrm{target}` that could z-level, z-star or z-tilde for now (and more general in the future).  The top layer would not be special.

Reply:
>  Ideally, there is a single function that is used for both ALE implementations, with and without VLR.

I think this should be a requirement (maybe to add to the requirement section even).  Redundant code would be quite risky in my opinion.

Reply:
@xylar, I believe the surface layer is currently treated differently for coastal wetting + drying (as I understand it, this is the only layer that can flood the coastal zone), but agree that we want the general formulation for VLR --- setting a target grid for all layers.

Reply:
I see, the coastal wetting-and-drying case is certainly one we should support but maybe not one that needs special attention in this design document, except to say that it would be another target grid in addition to z-star and z-tilde that we would want to support.

Reply:
> Since k is the vertical index, shouldn't this be z_1, z_2, etc. instead of z_k^1, z_k^2, etc?
This isn't the best notation choice. In general the superscript is the timestep marker. These numbers are just referring to temporary, intermediate grid choices, but they're maybe better expressed as updates to `z_k^{target}` prior to `h_k` actually being reassigned to be consistent with `z_k^{target}`.

Reply:
I guess I'm still not following but I agree that the numbers are not a good choice for provisional variables.  The POP convention was to use a `*` or some similar symbol to indicate a provisional variable.  Perhaps something like that would be more typical.